File: 1923-sentence-similarity-iii/1923-sentence-similarity-iii.cpp
class Solution {
public:
    bool areSentencesSimilar(string sentence1, string sentence2) {
        // Split the sentences into words
        vector<string> words1 = split(sentence1);
        vector<string> words2 = split(sentence2);
        
        // Ensure words1 is the longer sentence
        if (words1.size() < words2.size()) {
            swap(words1, words2);
        }
        
        int i = 0, j = 0;
        int n1 = words1.size(), n2 = words2.size();
        
        // Check for matching prefix
        while (i < n2 && words1[i] == words2[i]) {
            i++;
        }
        
        // Check for matching suffix
        while (j < n2 - i && words1[n1 - j - 1] == words2[n2 - j - 1]) {
            j++;
        }
        
        // The sentences are similar if all words from the smaller sentence are covered
        return i + j >= n2;
    }

private:
    // Helper function to split the sentence into words
    vector<string> split(const string& sentence) {
        vector<string> words;
        istringstream stream(sentence);
        string word;
        while (stream >> word) {
            words.push_back(word);
        }
        return words;
    }
};

File: 1923-sentence-similarity-iii/1923-sentence-similarity-iii.py
class Solution(object):
    def areSentencesSimilar(self, sentence1, sentence2):
        """
        :type sentence1: str
        :type sentence2: str
        :rtype: bool
        """
        # Split both sentences into words
        words1 = sentence1.split()
        words2 = sentence2.split()
        
        # Ensure words1 is the longer sentence
        if len(words1) < len(words2):
            words1, words2 = words2, words1
        
        # Two pointers: one from the start (i) and one from the end (j)
        i, j = 0, 0
        
        # Check for matching prefix
        while i < len(words2) and words1[i] == words2[i]:
            i += 1
        
        # Check for matching suffix
        while j < len(words2) - i and words1[-(j+1)] == words2[-(j+1)]:
            j += 1
        
        # The sentences are similar if all words from sentence2 are consumed
        return i + j >= len(words2)